toolchain/sem_ir/node.h

```diff
@@ -14,542 +14,118 @@
-struct TypedNode;
-
-// The standard structure for Node. This is trying to provide a minimal
-// amount of information for a node:
+// A type-erased representation of a SemIR node, that may be constructed from
+// the specific kinds of node defined in `typed_nodes.h`. This provides access
+// to common fields present on most or all kinds of nodes:
 //
 // - `parse_node` for error placement.
 // - `kind` for run-time logic when the input Kind is unknown.
 // - `type_id` for quick type checking.
-// - Up to two Kind-specific members.
 //
-// To create a specific kind of `Node`, use the appropriate `TypedNode`
-// constructor. A `TypedNode` implicitly converts to a `Node`.
+// In addition, any kind-specific data is stored and can be accessed by casting
+// to the specific kind of node:
 //
-// Given a `Node`, you may:
-//
-// - Access non-Kind-specific members like `Print`.
-// - Use `node.kind()` or `Is<Kind>` to determine what kind of node it is.
-// - Access Kind-specific members using `node.As<Kind>()`, which produces a
-//   `TypedNode` with type-specific members, including `parse_node` and
-//   `type_id` for nodes that have associated parse nodes and types.
-//   - Using the wrong kind in `node.As<Kind>()` is a programming error, and
-//     will CHECK-fail in debug modes (opt may too, but it's not an API
+// - Use `node.kind()` or `Is<TypedNode>` to determine what kind of node it is.
+// - Cast to a specific type using `node.As<TypedNode>()`
+//   - Using the wrong kind in `node.As<TypedNode>()` is a programming error,
+//     and will CHECK-fail in debug modes (opt may too, but it's not an API
 //     guarantee).
-// - Use `node.TryAs<Kind>()` to safely access type-specific node data where
-//   the node's kind is not known.
+// - Use `node.TryAs<TypedNode>()` to safely access type-specific node data
+//   where the node's kind is not known.
 class Node : public Printable<Node> {
  public:
-  template <NodeKind::RawEnumType Kind, typename Data>
+  template <typename TypedNode, typename Info = TypedNodeArgsInfo<TypedNode>>
   /*implicit*/
-  Node(TypedNode<Kind, Data> typed_node)
-      : Node(typed_node.parse_node_or_invalid(), NodeKind::Create(Kind),
-             typed_node.type_id_or_invalid(), typed_node.arg0_or_invalid(),
-             typed_node.arg1_or_invalid()) {}
+  Node(TypedNode typed_node)
+      : parse_node_(Parse::Node::Invalid),
+        kind_(TypedNode::Kind),
+        type_id_(TypeId::Invalid),
+        arg0_(NodeId::InvalidIndex),
+        arg1_(NodeId::InvalidIndex) {
+    if constexpr (HasParseNode<TypedNode>) {
+      parse_node_ = typed_node.parse_node;
+    }
+    if constexpr (HasTypeId<TypedNode>) {
+      type_id_ = typed_node.type_id;
+    }
+    if constexpr (Info::NumArgs > 0) {
+      auto tuple = StructReflection::AsTuple(typed_node);
+      arg0_ = ToRaw(std::get<Info::Arg0Field>(tuple));
+      if constexpr (Info::NumArgs > 1) {
+        arg1_ = ToRaw(std::get<Info::Arg0Field + 1>(tuple));
+      }
```

Comment:
Additional `Info` methods for getting the fields would be useful here too.

Reply:
Done.